feat: apkam-aware activate_cli

gkc opened the pull request:
**- What I did**
- feat: 'activate' CLI is now APKAM-aware, and supports 
  - onboarding (as before)
  - submitting enrollment requests
  - listing / approving / denying / revoking enrollment requests
  - generating one-time passcodes
  - setting semi-permanent passcode

**- How I did it**
- New code in src/cli : auth_cli.dart, auth_cli_args.dart
- Modified existing bin/activate_cli.dart so it uses auth_cli.main instead of activate_cli.main

**- How to verify it**
- See NoPorts e2e tests ([initial PR](https://github.com/atsign-foundation/noports/pull/1037)) which use this CLI extensively
- Some examples if you wish to test manually:
  - `dart bin/activate_cli.dart -h` to see overall usage info and list of available commands
  - `dart bin/activate_cli.dart <some command> -h` to see help and usage info for a specific command
  - `dart bin/activate_cli.dart otp -a @alice` to generate an OTP for example
  - `dart bin/activate_cli.dart interactive -a @alice` to run interactively

**-Additional context**
- TODO but will be in follow-up PR as it's not on critical path right now: Add code which allows user to 'resume' an enrollment similar to what has been done in at_client_mobile - basically if the CLI is killed while waiting for its enrollment request to be approved or denied, then allow the exact same command to be re-run, but instead of submitting a new enrollment request it finds the details of the previous one and then goes straight to attempting PKAM auth

gkc committed:
chore:  AtOnboardingPreference: remove unnecessary overrides of AtCientPreference fields

gkc committed:
feat: new apkam-aware activate / onboarding CLI: interim commit

gkc committed:
Merge remote-tracking branch 'origin/trunk' into gkc/apkam-aware-cli

gkc committed:
Merge remote-tracking branch 'origin/trunk' into gkc/apkam-aware-cli

# Conflicts:
#	packages/at_onboarding_cli/bin/activate_cli.dart
#	packages/at_onboarding_cli/pubspec.yaml

gkc committed:
feat: add Duration? retryInterval param to AtOnboardingService.enroll; deprecate the int? pkamRetryIntervalMins parameter

gkc committed:
feat: new apkam-aware cli for onboarding / enrollment / enrollment management. Interim commit: mostly structural changes for testability, readability etc

gkc committed:
fix: Fixed async-await bug in at_onboarding_service_impl. Added more lint rules to analysis_options.yaml including the unawaited futures rule so we don't have this happen anywhere else.

gkc committed:
feat: new at_activate cli working fully; needs a couple of changes to go to prod swarms before can be used for prod

gkc committed:
use new enroll:fetch:$id to fetch a specific enrollment

gkc committed:
Merge remote-tracking branch 'origin/trunk' into gkc/apkam-aware-cli

gkc committed:
Merge remote-tracking branch 'origin/trunk' into gkc/apkam-aware-cli

gkc committed:
docs: update CHANGELOG

gkc committed:
Merge remote-tracking branch 'origin/trunk' into gkc/apkam-aware-cli

gkc committed:
feat: at_onboarding_cli: Added regex support for list / approve / deny / revoke in auth_cli

gkc committed:
build: at_onboarding_cli: pubspec and CHANGELOG for 1.5.0

gkc committed:
docs: at_onboarding_cli: tweaked pub description

gkc committed:
feat: at_onboarding_cli: auth_cli throw ArgumentError when appropriate

gkc committed:
fix: auth_cli: write generated OTP to stdout

gkc committed:
build: at_onboarding_cli: upgrade at_cli_commons dependency

gkc committed:
feat: auth_cli: add maxConnectAttempts param to auth_cli

gkc committed:
build: using published version 1.1.0 of at_cli_commons

gkc committed:
test: modify cli functional tests pubspec to use at_onboarding_cli from pub.dev rather than local path. This is slightly inconvenient but we have a circular dependency now between at_onboarding_cli and at_cli_commons.

gkc commented on packages/at_onboarding_cli/lib/src/onboard/at_onboarding_service_impl.dart:
```diff
@@ -131,15 +131,21 @@ class AtOnboardingServiceImpl implements AtOnboardingService {
-    final Duration retryInterval = Duration(minutes: pkamRetryIntervalMins);
+    retryInterval ??= Duration(
+        minutes: pkamRetryIntervalMins ??
+            atOnboardingPreference.apkamAuthRetryDurationMins);
```
added retryInterval as a Duration, so calling code has more control

gkc commented on packages/at_onboarding_cli/lib/src/onboard/at_onboarding_service_impl.dart:
```diff
@@ -160,11 +166,11 @@ class AtOnboardingServiceImpl implements AtOnboardingService {
 
     // Pkam auth will be attempted asynchronously until enrollment is approved/denied
-    _attemptPkamAuthAsync(
+    await _attemptPkamAuthAsync(
```
added missing await

gkc commented on packages/at_onboarding_cli/lib/src/onboard/at_onboarding_service_impl.dart:
```diff
@@ -174,12 +180,13 @@ class AtOnboardingServiceImpl implements AtOnboardingService {
   }
 
-  void _listenToPkamSuccessStream(
+  Future<void> _listenToPkamSuccessStream (
```
Made _listenToPkamSuccessStream return a Future which is only completed when we've actually processed the success

gkc commented on packages/at_onboarding_cli/lib/src/util/print_full_parser_usage.dart:
```diff
@@ -0,0 +1,52 @@
+
+import 'package:args/args.dart';
+
+extension PrintAllArgParserUsage on ArgParser {
```
A little extension to help with pretty printing of the auth_cli's ArgParsers' usage when there are subcommands involved

gkc commented on packages/at_onboarding_cli/lib/src/cli/auth_cli.dart:
```diff
@@ -0,0 +1,661 @@
+  final first = arguments.first;
+  if (first.startsWith('-') && first != '-h' && first != '--help') {
+    // no command found ... legacy ... insert 'onboard' as the command
+    arguments = ['onboard', ...arguments];
```
This is the only tricky bit of logic to be aware of, to allow us to most easily provide backwards compatibility

murali-shris commented:
@gkc While running enroll using activate_cli, I am getting this
```
murali@Muralidharans-MacBook-Pro at_onboarding_cli % dart bin/activate_cli.dart enroll -s 1P14P1  -r vip.ve.atsign.zone -p buzz -d pixel -n “ "buzz:rw” -a @alice🛠 -k /Users/murali/.atsign/keys/@alice🛠_buzz.atKeys
dquote> 
```

murali-shris commented:
@gkc 
```
murali@Muralidharans-MacBook-Pro at_onboarding_cli % dart bin/activate_cli.dart enroll -s 5KSNW3  -r vip.ve.atsign.zone -p buzz -d pixel -n “buzz:rw” -a @alice🛠 -k /Users/murali/.atsign/keys/@alice🛠_buzz.atKeys
```
Can we log the enrollment response from server after running the above command. I had to copy the enrollmentId from the server logs and pass it to the approve command.

murali-shris commented:
With dart 3.3.4, I am getting this warning every time I run activate_cli
```
murali@Muralidharans-MacBook-Pro at_onboarding_cli % dart bin/activate_cli.dart enroll -s 5KSNW3  -r vip.ve.atsign.zone -p buzz -d pixel -n “buzz:rw” -a @alice🛠 -k /Users/murali/.atsign/keys/@alice🛠_buzz.atKeys
../../../../../.pub-cache/hosted/pub.dev/pointycastle-3.9.0/lib/macs/cbc_block_cipher_mac.dart:103:46: Warning: Operand of null-aware operation '!' has type 'Padding' which excludes null.
 - 'Padding' is from 'package:pointycastle/api.dart' ('../../../../../.pub-cache/hosted/pub.dev/pointycastle-3.9.0/lib/api.dart').
    var paddingName = _padding != null ? '/${_padding!.algorithmName}' : '';
                                             ^
../../../../../.pub-cache/hosted/pub.dev/pointycastle-3.9.0/lib/macs/cbc_block_cipher_mac.dart:203:7: Warning: Operand of null-aware operation '!' has type 'Padding' which excludes null.
 - 'Padding' is from 'package:pointycastle/api.dart' ('../../../../../.pub-cache/hosted/pub.dev/pointycastle-3.9.0/lib/api.dart').
      _padding!.addPadding(_buf, _bufOff);
      ^
../../../../../.pub-cache/hosted/pub.dev/pointycastle-3.9.0/lib/signers/ecdsa_signer.dart:94:41: Warning: Operand of null-aware operation '!' has type 'Mac' which excludes null.
 - 'Mac' is from 'package:pointycastle/api.dart' ('../../../../../.pub-cache/hosted/pub.dev/pointycastle-3.9.0/lib/api.dart').
      kCalculator = _RFC6979KCalculator(_kMac!, n, _pvkey!.d!, message);
                                        ^
../../../../../.pub-cache/hosted/pub.dev/pointycastle-3.9.0/lib/signers/ecdsa_signer.dart:166:7: Warning: Operand of null-aware operation '!' has type 'Digest' which excludes null.
 - 'Digest' is from 'package:pointycastle/api.dart' ('../../../../../.pub-cache/hosted/pub.dev/pointycastle-3.9.0/lib/api.dart').
      _digest!.reset();
      ^
../../../../../.pub-cache/hosted/pub.dev/pointycastle-3.9.0/lib/signers/ecdsa_signer.dart:167:14: Warning: Operand of null-aware operation '!' has type 'Digest' which excludes null.
 - 'Digest' is from 'package:pointycastle/api.dart' ('../../../../../.pub-cache/hosted/pub.dev/pointycastle-3.9.0/lib/api.dart').
      return _digest!.process(message);

```

gkc commented:
> With dart 3.3.4, I am getting this warning every time I run activate_cli
> 
> ```
> murali@Muralidharans-MacBook-Pro at_onboarding_cli % dart bin/activate_cli.dart enroll -s 5KSNW3  -r vip.ve.atsign.zone -p buzz -d pixel -n “buzz:rw” -a @alice🛠 -k /Users/murali/.atsign/keys/@alice🛠_buzz.atKeys
> ../../../../../.pub-cache/hosted/pub.dev/pointycastle-3.9.0/lib/macs/cbc_block_cipher_mac.dart:103:46: Warning: Operand of null-aware operation '!' has type 'Padding' which excludes null.
>  - 'Padding' is from 'package:pointycastle/api.dart' ('../../../../../.pub-cache/hosted/pub.dev/pointycastle-3.9.0/lib/api.dart').
>     var paddingName = _padding != null ? '/${_padding!.algorithmName}' : '';
>                                              ^
> ../../../../../.pub-cache/hosted/pub.dev/pointycastle-3.9.0/lib/macs/cbc_block_cipher_mac.dart:203:7: Warning: Operand of null-aware operation '!' has type 'Padding' which excludes null.
>  - 'Padding' is from 'package:pointycastle/api.dart' ('../../../../../.pub-cache/hosted/pub.dev/pointycastle-3.9.0/lib/api.dart').
>       _padding!.addPadding(_buf, _bufOff);
>       ^
> ../../../../../.pub-cache/hosted/pub.dev/pointycastle-3.9.0/lib/signers/ecdsa_signer.dart:94:41: Warning: Operand of null-aware operation '!' has type 'Mac' which excludes null.
>  - 'Mac' is from 'package:pointycastle/api.dart' ('../../../../../.pub-cache/hosted/pub.dev/pointycastle-3.9.0/lib/api.dart').
>       kCalculator = _RFC6979KCalculator(_kMac!, n, _pvkey!.d!, message);
>                                         ^
> ../../../../../.pub-cache/hosted/pub.dev/pointycastle-3.9.0/lib/signers/ecdsa_signer.dart:166:7: Warning: Operand of null-aware operation '!' has type 'Digest' which excludes null.
>  - 'Digest' is from 'package:pointycastle/api.dart' ('../../../../../.pub-cache/hosted/pub.dev/pointycastle-3.9.0/lib/api.dart').
>       _digest!.reset();
>       ^
> ../../../../../.pub-cache/hosted/pub.dev/pointycastle-3.9.0/lib/signers/ecdsa_signer.dart:167:14: Warning: Operand of null-aware operation '!' has type 'Digest' which excludes null.
>  - 'Digest' is from 'package:pointycastle/api.dart' ('../../../../../.pub-cache/hosted/pub.dev/pointycastle-3.9.0/lib/api.dart').
>       return _digest!.process(message);
> ```

Should not occur if you do `dart pub upgrade` to pull in the latest pointycastle

gkc commented:
> @gkc
> 
> ```
> murali@Muralidharans-MacBook-Pro at_onboarding_cli % dart bin/activate_cli.dart enroll -s 5KSNW3  -r vip.ve.atsign.zone -p buzz -d pixel -n “buzz:rw” -a @alice🛠 -k /Users/murali/.atsign/keys/@alice🛠_buzz.atKeys
> ```
> 
> Can we log the enrollment response from server after running the above command. I had to copy the enrollmentId from the server logs and pass it to the approve command.

Done in latest commit

gkc committed:
at_onboarding_cli: Improve the terminal output from activate_cli. Also some refactoring
- Deprecated `AtOnboardingPreference.apkamAuthRetryDurationMins`
- Added three new methods to AtOnboardingService - sendEnrollRequest, awaitApproval and createAtKeysFile. Extracted those methods from AtOnboardingServiceImpl.enroll so it now just calls those three methods. The refactoring provides more control to application code.
- For readability, refactored the code related to pkam auth in the context of enrollment success checking

gkc committed:
chore: run dart format

gkc commented:
@murali-shris please can you re-review?
